File: apps/php/design-pattern/src/TemplateMethod/Report.php
<?php

declare(strict_types=1);

namespace DesignPattern\TemplateMethod;

abstract class Report
{
    protected string $title;
    protected array $text;

    public function __construct(string $title = '月次報告', array $text = ['順調', '最高の調子'])
    {
        $this->title = $title;
        $this->text = $text;
    }

    /** テンプレートメソッド: レポート出力の骨格 */
    public function outputReport(): string
    {
        $lines = [];
        $lines = array_merge($lines, $this->outputStart());
        $lines = array_merge($lines, $this->outputHead());
        $lines = array_merge($lines, $this->outputBodyStart());
        $lines = array_merge($lines, $this->outputBody());
        $lines = array_merge($lines, $this->outputBodyEnd());
        $lines = array_merge($lines, $this->outputEnd());
        return implode("\n", $lines);
    }

    protected function outputBody(): array
    {
        $lines = [];
        foreach ($this->text as $line) {
            $lines = array_merge($lines, $this->outputLine($line));
        }
        return $lines;
    }

    /** フックメソッド（デフォルトは何もしない） */
    protected function outputStart(): array
    {
        return [];
    }

    protected function outputHead(): array
    {
        return $this->outputLine($this->title);
    }

    protected function outputBodyStart(): array
    {
        return [];
    }

    /** 抽象メソッド */
    abstract protected function outputLine(string $line): array;

    protected function outputBodyEnd(): array
    {
        return [];
    }

    protected function outputEnd(): array
    {
        return [];
    }
}

File: apps/java/design-pattern/src/main/java/pattern/templatemethod/Report.java
package pattern.templatemethod;

import java.util.List;

/**
 * 基底レポートクラス（Template Method パターン）
 *
 * テンプレートメソッド outputReport() が処理の骨格を定義し、
 * サブクラスが各ステップ（フックメソッド）をオーバーライドする。
 */
public abstract class Report {

    protected final String title = "月次報告";
    protected final List<String> text = List.of("順調", "最高の調子");

    /**
     * テンプレートメソッド: レポート出力の骨格
     */
    public String outputReport() {
        StringBuilder sb = new StringBuilder();
        outputStart(sb);
        outputHead(sb);
        outputBodyStart(sb);
        outputBody(sb);
        outputBodyEnd(sb);
        outputEnd(sb);
        return sb.toString();
    }

    protected void outputBody(StringBuilder sb) {
        for (String line : text) {
            outputLine(sb, line);
        }
    }

    // フックメソッド（デフォルトは何もしない）
    protected void outputStart(StringBuilder sb) {}

    protected void outputHead(StringBuilder sb) {
        outputLine(sb, title);
    }

    protected void outputBodyStart(StringBuilder sb) {}

    /**
     * 抽象メソッド: サブクラスでオーバーライド必須
     */
    protected abstract void outputLine(StringBuilder sb, String line);

    protected void outputBodyEnd(StringBuilder sb) {}

    protected void outputEnd(StringBuilder sb) {}
}
